connmgr/connmanager.go

```diff
@@ -0,0 +1,229 @@
+				return
+			}
+			numPeers := len(seedpeers)
+
+			log.Infof("%d addresses found from DNS seed %s", numPeers, seeder)
+
+			if numPeers == 0 {
+				return
+			}
+			addresses := make([]*wire.NetAddress, len(seedpeers))
+			// if this errors then we have *real* problems
+			intPort, _ := strconv.Atoi(ChainParams.DefaultPort)
+			for i, peer := range seedpeers {
+				addresses[i] = new(wire.NetAddress)
+				addresses[i].SetAddress(peer, uint16(intPort))
+				// bitcoind seeds with addresses from
+				// a time randomly selected between 3
+				// and 7 days ago.
+				addresses[i].Timestamp = time.Now().Add(-1 *
+					time.Second * time.Duration(secondsIn3Days+
+					randSource.Int31n(secondsIn4Days)))
+			}
+
+			// Bitcoind uses a lookup of the dns seeder here. This
+			// is rather strange since the values looked up by the
+			// DNS seed lookups will vary quite a lot.
+			// to replicate this behaviour we put all addresses as
+			// having come from the first one.
+			cm.amgr.AddAddresses(addresses, addresses[0])
+		}(seeder)
+	}
+}
+
+// connectionManager manages a fixed number of outbound connections.
+func (cm *ConnManager) connectionManager() {
+	time.AfterFunc(10*time.Second, func() { cm.wakeup <- struct{}{} })
+out:
+	for {
+		select {
+		case <-cm.wakeup:
+			if len(cm.connections) < MaxOutboundPeers {
+				addr := cm.amgr.GetAddress("any")
+				if addr == nil {
+					break
+				}
+				key := addrmgr.GroupKey(addr.NetAddress())
+				// Address will not be invalid, local or unroutable
+				// because addrmanager rejects those on addition.
+				// Just check that we don't already have an address
+				// in the same group so that we are not connecting
+				// to the same network segment at the expense of
+				// others.
+				if cm.outboundGroups[key] != 0 {
+					break
+				}
+
+				addrStr := addrmgr.NetAddressKey(addr.NetAddress())
+
+				// allow nondefault ports after 50 failed tries.
+				if fmt.Sprintf("%d", addr.NetAddress().Port) !=
+					ChainParams.DefaultPort {
+					continue
+				}
+
+				cr := cm.Connect(addrStr, false)
+				go cm.NewConnectionHandler(cr)
+			}
+			time.AfterFunc(10*time.Second, func() { cm.wakeup <- struct{}{} })
+
+		case <-cm.quit:
+			break out
+		}
+	}
+	cm.wg.Done()
+}
+
+// connectionHandler is a service that monitors requests for new
+// connections or closed connections and maps addresses to their
+// respective connections.
+func (cm *ConnManager) connectionHandler() {
+out:
+	for {
+		select {
+		case cr := <-cm.newConnections:
+			cm.connections[cr.Addr] = cr.Conn
+
+		case addr := <-cm.closeConnections:
+			c := cm.connections[addr]
+			err := c.Close()
+			if err != nil {
+				log.Infof("Error closing connection %s: %v", addr, err)
+			}
+			delete(cm.connections, addr)
+
+		case <-cm.quit:
+			break out
+		}
+	}
+	cm.wg.Done()
+}
+
+// Start launches the connection manager.
+func (cm *ConnManager) Start() {
+	cm.seedFromDNS(ChainParams.DNSSeeds)
```

Comment:
As mentioned elsewhere, the `ChainParams` should come from a `Config` on the `ConnManager` the same as other packages handle it.

Also, I don't see any reason to pass this variable to the `cm.seedFromDNS` function since it has access to the `ConnManager` state and it's not called from anywhere else with different values.